salt/states/pkg.py

```diff
@@ -669,6 +672,9 @@ def _find_install_targets(name=None,
     failed_verify = False
     for key, val in six.iteritems(desired):
         cver = cur_pkgs.get(key, [])
+        if resolve_capabilities and not cver and key in cur_prov:
+            cver = cur_pkgs.get(cur_prov.get(key)[0], [])
```

Comment:
This is quite evil coding. :wink: Everywhere below I would call rather `whatever[:1]` instead of `whatever[0]`. The `[:-1]` construction won't crash if `cur_prov.get(key)` returns an empty array (and you still expecting an array, never `None`)! Also I would not assign `cver` twice. So I would write this a bit shorter:
```python
# desired.items() also OK, except doesn't return an iterator as Python 3.
for key, val in six.iteritems(desired):
    cver = cur_pkgs.get(key, cur_pkgs.get(cur_prov.get(key, [])[:1]) if resolve_capabilities else [])

Reply:
I don't think that your code is doing the same as the code above :-)
I think we need to discuss this in person.

Reply:
I wanted to try this but it does not work:

    TypeError: unhashable type: 'list'

Reply:
Ah, my bad. The [:1] is still `[element]`, not `element`. Well, use `[0]` for now, if you are 100% sure it comes always an array with at least one element in it.

Reply:
Right. Then just `cur_pkgs.get(cur_prov.get(key, [None])[0]`, if you are 100% sure that `key` once found in `cur_prov` contains non-empty array.

Reply:
OK, here. You have it:
```python
for key, val in desired.items():                                                                                
    cver = cur_pkgs.get(key, cur_pkgs.get(cur_prov.get(key)[0], []) if resolve_capabilities else [])
```

Reply:
No, does not work in all cases: 

    TypeError: 'NoneType' object has no attribute '__getitem__'

Reply:
Which means either `cur_pkgs` or `cur_prov` is not a dictionary.